### DISCUSSION TITLE
[PROPOSAL] Lazy initialization of model

### DISCUSSION THREAD

**Discussion by ver217** [hpcaitech/ColossalAI#3124] [Development | Core]
## What are LazyTensor and LazyInit

LazyTensor allows DL framework (PyTorch) to execute operations lazily, by storing all operations related to it and reruning them when it's required to be materialized.

LazyInit defers model initialiazation and it's based on LazyTensor.

## Why we need implement new LazyInit

ColossalAI actually has similar feature now: [ColoInitContext](https://github.com/hpcaitech/ColossalAI/blob/68577fbc4399b0e8333ad958959ac09e5c54033d/colossalai/utils/model/colo_init_context.py#L69). It hijacks the `__init__()` method of each module and shard each layer with preset sharding strategy.

![old lazy init](https://user-images.githubusercontent.com/23111350/224652072-93a525fd-fbc6-49d1-a105-84f10d4c0841.png)

This can work well when sharding strategy is known before model initialization. If sharding strategy is generated from static analysis about model, this method won't work anymore.

So we need initialize model tensors using meta tensor and do static analysis to get sharding strategy. And then materialize each tensor and apply the sharding strategy. The static analysis can be omitted if the sharding strategy is known in advance.

![new lazy init](https://user-images.githubusercontent.com/23111350/224655553-f7009b87-a266-40cb-bad0-64dca57661cb.png)

A possible initialization process:

![init process](https://user-images.githubusercontent.com/23111350/224653657-7948559d-1112-4d8b-9ecd-f776487d3d12.png)

Actually [torchdistx](https://github.com/pytorch/torchdistx) has implemented similar features named `FakeTensor` and `deferred_init`.

Why do we need implement a new one?

1. It is hard to verify the correctness of initialization. There may be random initialization and we have to control the seed or random number generator state.
2. This may be combined with `DTensor`. But `torchdistx` is a kind of blackbox to us.

## Method

We have experimental code about [lazy init](https://github.com/hpcaitech/ColossalAI/blob/main/colossalai/utils/model/experimental.py). Thank @super-dainiu .

We can start this work by this file. We implement a LazyTensor class which tracks all OPs and rerun them when materializing.

A possible class definition of LazyTensor:

```python
class LazyTensor(torch.Tensor):
    _meta_data: Optional[MetaTensor] = None    # shape, dtype, device
    _pre_op_fn: Callable[['LazyTensor'], None] = lambda *args: None  # reserverd for verification
    _factory_method: Callable[[Any], Tensor]]
    _op_buffer = []
    _materialized_data: Optional[torch.Tensor] = None

    def materialize(self) -> torch.Tensor:
        pass

    def clean(self) -> None:
        pass
```

A possible usage of LazyInit:

```python
ctx = LazyInitContext()
with ctx:
    model = GPT()

# if no need to distribute model
ctx.materialize(model)

# if need distribute model
   
sharding_strategy = symbolic_trace(model, ...)
ctx.distribute(model, sharding_strategy )
```

`LazyInitContext.materialize()` and `LazyInitContext.distribute()` are static methods and may be replaced with downstream model wrappers.

## Limitations

To keep the implementation simple, we have some trade-off when designing.

We cannot ensure the lazy intialized model is the same as standard intialized model, but we can ensure its parameters are initialized from the same distribution.

There are some cases we cannot ensure the correctness:

- modify origin tensor after cloning. E.g. 
  ```python
  b = a.clone()
  a.add_(1)
  ```
- modify origin tensor after tolist. E.g.
  ```python
  b = a.tolist()
  a.add_(1)
  ```
As we don't track tensor's slice relationship, there are also some cases lazy execution won't work and tensors may be early materialized.

- get slice. E.g.
  ```python
  b = a[:, 2:]
  
  # modify slice directly won't trigger early materialization
  a[:, 2:] = 1.0
  ```
- split. E.g.
  ```python
  chunks = a.split(3)
  ```

## How to verify

To verify the correctness, we have to control the random seed. We can implement a utility tensor class:

```python
class MyTensor(Tensor):
    def __new__(cls, func, *args, dtype=None, device=None, **kwargs) -> 'MyTensor':
        set_seed(42)
        data = func(*args, dtype=dtype, device=device, **kwargs)
        return Tensor._make_subclass(cls, data, require_grad=data.requires_grad)

    @classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        set_seed(42)
        return super().__torch_function__(func, types, args, kwargs)
```

Thus, random states are same before each OP.

For `LazyTensor`, we reserve a hook to control random seed before executing each OP.

By doing this, we can simply compare the state dict or forward result to verify the correctness of initialization.

## Possible Roadmap

- [x] Implement a single-process version of lazy init.
- [x] Verify correctness on many models.
- [x] Combine with DTensor.
- [x] Verify on distributed training cases.

**Comment by Muggle666** [comment]
How to load pretrained weight in this context? Could you please provide an example?

**Reply by ver217** [reply]
We should not load pretrained weight in this context. If so, it's not lazy. We should first lazy initialize model and then load checkpoint. Here is an example (psuedo-code):

```python
with LazyInitContext():
    model = GPT2LMHeadModel(config)

optimizer = ...
lr_scheduler = ...
dataloader = ...
model, optimizer, lr_scheduler, dataloader = booster.boost(model, optimizer, lr_scheduler, dataloader)

booster.load_model(model, pretrained_path)
```

`Booster` supports loading huggingface/transformers-fashion pretrained checkpoint, but it must be local.
